libs/redis/langchain_redis/vectorstores.py

```diff
@@ -933,200 +1011,78 @@ def similarity_search_with_score_by_vector(
 
             if with_vectors:
                 return_fields.append(self.config.embedding_field)
-        else:
-            return_fields = []
 
         if distance_threshold is None:
-            results = self._index.query(
-                VectorQuery(
-                    vector=embedding,
-                    vector_field_name=self.config.embedding_field,
-                    return_fields=return_fields,
-                    num_results=k,
-                    filter_expression=filter,
-                    sort_by=sort_by,
-                )
+            query = VectorQuery(
+                vector=embedding,
+                vector_field_name=self.config.embedding_field,
+                return_fields=return_fields,
+                num_results=k,
+                filter_expression=filter,
+                sort_by=sort_by,
             )
         else:
-            results = self._index.query(
-                RangeQuery(
-                    vector=embedding,
-                    vector_field_name=self.config.embedding_field,
-                    return_fields=return_fields,
-                    num_results=k,
-                    filter_expression=filter,
-                    sort_by=sort_by,
-                    distance_threshold=distance_threshold,
-                )
+            query = RangeQuery(
+                vector=embedding,
+                vector_field_name=self.config.embedding_field,
+                return_fields=return_fields,
+                num_results=k,
+                filter_expression=filter,
+                sort_by=sort_by,
+                distance_threshold=distance_threshold,
             )
 
         if not return_all:
             if with_vectors:
-                # Extract the document ids
-                doc_ids = [doc["id"] for doc in results]
-
-                # Retrieve the documents from the storage
-                docs_from_storage = self._index._storage.get(
-                    self._index.client, doc_ids
+                query.return_field(
```

Comment:
also, are we sure this doesn't conflict with line 1012?

Reply:
It does not one creates the metadata object that is received later but they don't both touch the query. I believe the tests would fail otherwise as well.

I still don't 100% love this branching logic and had started a larger refactor but cut bait for scope creep

Reply:
good call for today